### PR TITLE
[IMP] base: Document layout preview improvements 

### PR DESCRIPTION
**Pull request by SimonGenin**
[IMP] base: Document layout preview improvements

The document layout preview is a complete defferent simplified template
with its own css that replicates at best the different styles.
It does not have the external layout features and lack of fidelity.

The new preview actually use the real documents templates and put the
result in an iframe. It now has a high fidelity, though not perfect.

The goal is for a better onboarding, where clients see easely how
documents will look if they had an app to generate them. Of course, the
data on the document is a false invoice.

Migration: [https://github.com/odoo/upgrade/pull/1675](https://github.com/odoo/upgrade/pull/1675)

### CODE REVIEW

File: odoo/addons/base/models/ir_ui_view.py [hunk 1]
```diff
@@ -1688,7 +1688,7 @@ def get_view_xmlid(self):
     @api.model
     def render_public_asset(self, template, values=None):
         template = self.sudo().browse(self.get_view_id(template))
-        template._check_view_access()
+        # template._check_view_access()
```
Comment (Arcasias) [inline odoo/addons/base/models/ir_ui_view.py:1691]: Remove if not useful

File: addons/web/models/base_document_layout.py [hunk 2]
```diff
@@ -58,13 +85,17 @@ def _compute_logo_colors(self):
                 wizard_for_image = wizard
             wizard.logo_primary_color, wizard.logo_secondary_color = wizard_for_image._parse_logo_colors()
 
-    @api.depends('report_layout_id', 'logo', 'font', 'primary_color', 'secondary_color')
+    @api.depends('report_layout_id', 'logo', 'font', 'primary_color', 'secondary_color', 'report_header', 'report_footer', 'external_report_layout_id')
```
Comment (Arcasias) [inline addons/web/models/base_document_layout.py:88]: Not sure about `'external_report_layout_id'`, could you explain?

File: addons/web/views/report_templates.xml [hunk 3]
```diff
@@ -66,13 +61,244 @@
...
         </t>
     </template>
 
+    <template id="html_preview_container">
+        <t t-set="body_classname" t-value="'container'"/>
+            <t t-call="web.report_preview_layout">
+                <t t-raw="0"/>
+            </t>
+    </template>
+
```
Comment (Arcasias) [inline addons/web/views/report_templates.xml:114]: There are quite a lot of empty lines in this diff; in general we only use them to separate different templates

File: addons/web/views/report_templates.xml [hunk 4]
```diff
@@ -66,13 +61,244 @@
...
     <template id="html_container">
         <t t-set="body_classname" t-value="'container'"/>
         <t t-call="web.report_layout">
             <t t-raw="0"/>
         </t>
     </template>
 
+    <template id="html_preview_container">
+        <t t-set="body_classname" t-value="'container'"/>
+            <t t-call="web.report_preview_layout">
```
Comment (Arcasias) [inline addons/web/views/report_templates.xml:110]: Indentation

File: addons/web/views/report_templates.xml [hunk 5]
```diff
@@ -424,7 +650,12 @@
             <t t-set="primary" t-value="company.primary_color or 'black'" />
             <t t-set="secondary" t-value="company.secondary_color or 'black'" />
             <t t-set="layout" t-value="company.external_report_layout_id.key or 'web.external_layout_standard'" />
-            .o_company_<t t-esc='company.id'/>_layout {
+            <t t-if="PREVIEW_ID">
+                 .o_company___PREVIEW_ID___layout {
+            </t>
+            <t t-if="not PREVIEW_ID" >
```
Comment (Arcasias) [inline addons/web/views/report_templates.xml:656]:
```suggestion
            <t t-else="" >
```

**Review by Arcasias — CHANGES_REQUESTED** [review]
Great job overall, the preview seems to be working fine!
A few remarks additionnally to the review comments:

It would be nice if you could split the commit in at least 2 parts:

1. A commit to move all the code related to document_layout from base to web
2. Another one with the actual changes to the preview system

This would be a lot easier to read and make it possible to isolate problems in the future if they originate from this PR.

Also for some reasons I can't use the "Reset to logo colors" button anymore. I have yet to find why but it could be related to your changes.

File: addons/web/static/src/js/fields/special_fields.js [hunk 6]
```diff
@@ -192,9 +196,84 @@ var FieldReportLayout = relational_fields.FieldMany2One.extend({
...
+        let resolver;
+        $(iframe).data('ready', new Promise(function (resolve) {
+            resolver = resolve
+        }));
+
+
+        /**
+         * Certain browser don't trigger onload events of iframe for particular cases.
+         * In our case, chrome and safari could be problematic (yet not on all machines).
+         * This rather unorthodox solution replace the onload event handler.
```
Comment (Arcasias) [inline addons/web/static/src/js/fields/special_fields.js:237]: What do you mean by "yet not on all machines"? Is this more than a browser compatiblity problem? Currently the JS Framework supports all evergreen browsers (mainly if not only: Chrome & Chromium, Opera, Safari, Firefox and Edge + their mobile equivalents).

Now if it is machine-specific I would like to know the details because I think this is a rather huge trade-off.

File: addons/web/static/src/scss/base_document_layout.scss [hunk 7]
```diff
@@ -1,150 +1,38 @@
...
-        flex: 4;
-        border: 1px solid;
-        border-color: gray;
-        margin-right: 12px;
-        padding: 12px;
-        display: flex;
-        min-height: 420px;
-        flex-direction: column;
-    }
+#o_preview_iframe_wrapper {
```
Comment (Arcasias) [inline addons/web/static/src/scss/base_document_layout.scss:8]: The style definition of the IframeWrapper widget should not be here. You can move it to the `special_fields.scss` file that I mentionned earlier

**Review by Arcasias — COMMENTED** [review]
Leaving a few more remarks, sorry for the double review

File: addons/web/static/src/js/fields/special_fields.js (hunk 6 above)
Comment (SimonGenin) [reply addons/web/static/src/js/fields/special_fields.js:237]: So for what I know: It works on chrome on my mac. It doesn't on safari. When Nathan tried it on his laptop with chrome, it didn't worked. That's why I said "not on all machines". The truth is probably different but those iframes are touchy. Also, from the documentation and comments on iframe.onload, it is quite unreliable. It also can sometimes draw the report once, meaning the onload was properly fired, but won't be fired anymore for changes. From looking on the internet, depending on the browser, there are also different behaviors that can fire the event twice. And finally, still depending on the browser, it will behave differently depending on the attributes of the iframe. So I got a bit overwhelmed, tried a lot of things but that's currently my safest solution.

File: addons/web/static/src/scss/base_document_layout.scss [hunk 8]
```diff
@@ -1,150 +1,38 @@
-.o_document_layout .o_group {
```
[the remark this answers is not in the document]
Comment (SimonGenin) [reply addons/web/static/src/scss/base_document_layout.scss:1]: oups

File: addons/web/static/src/js/fields/special_fields.js [hunk 9]
```diff
@@ -192,9 +196,84 @@ var FieldReportLayout = relational_fields.FieldMany2One.extend({
...
+
+    _render: function () {
+
+        this.$el.empty();
+        this.$el.attr('id', 'o_preview_iframe_wrapper')
+
+        const $spinner = $('<i class="fa fa-spinner fa-spin fa-4x" id="o_iframe_wrapper_spinner" style="color: #875a7b; z-index: 2"></i>')[0]
+        this.$el[0].appendChild($spinner)
+
+        const html_code = this.value
```
[the remark this answers is not in the document]
Comment (SimonGenin) [reply addons/web/static/src/js/fields/special_fields.js:220]: I didn't know and my linter is not properly configured 🤡

File: addons/web/static/src/js/fields/special_fields.js [hunk 10]
```diff
@@ -192,9 +196,84 @@ var FieldReportLayout = relational_fields.FieldMany2One.extend({
...
+     * stored here: addons/web/static/src/scss/base_document_layout.scss and contains some style linked to this
+     * widget. Proceed with caution not to break the document layout preview.
+     */
+
+    _render: function () {
+
+        this.$el.empty();
+        this.$el.attr('id', 'o_preview_iframe_wrapper')
+
+        const $spinner = $('<i class="fa fa-spinner fa-spin fa-4x" id="o_iframe_wrapper_spinner" style="color: #875a7b; z-index: 2"></i>')[0]
```
[the remark this answers is not in the document]
Comment (SimonGenin) [reply addons/web/static/src/js/fields/special_fields.js:217]: @Arcasias Can use elaborate on "html ids is dangerous" ? I once read it could hurt performances somehow, but dangerous ? How ?

File: addons/web/models/base_document_layout.py (hunk 2 above)
Comment (SimonGenin) [reply addons/web/models/base_document_layout.py:88]: Can be removed, left over from the start of the task when I was trying to understand the code.

File: odoo/addons/base/models/ir_ui_view.py (hunk 1 above)
Comment (SimonGenin) [reply odoo/addons/base/models/ir_ui_view.py:1691]: ho that's dangerous, it was something I removed with Lucas cause we were denied access to assets. Only for tests purposes, I need to add the line back.

File: addons/web/static/src/js/fields/special_fields.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (Arcasias) [reply addons/web/static/src/js/fields/special_fields.js:217]: I don't know about performances, just that when using ids you have to be sure to use a unique string. When doing so, we usually use a prefix and a unique ID set on the component/widget (see the [Owl Datepicker](https://github.com/odoo/odoo/blob/master/addons/web/static/src/js/components/datepicker.js) for example). A typical use case is when using input elements coupled with a label having a `for` attribute.

So perhaps not "dangerous", but it can quickly cause inconsistencies if you're not careful enough.

File: addons/web/static/src/js/fields/special_fields.js (hunk 6 above)
Comment (SimonGenin) [reply addons/web/static/src/js/fields/special_fields.js:237]: @Arcasias

Comment (Arcasias) [reply addons/web/static/src/js/fields/special_fields.js:237]: Fair enough, let's keep it that way then.

**Comment by SimonGenin** [comment]
@Arcasias One note though, I don't know how to create two commits from the one I have. Any tips ?

**Comment by Arcasias** [comment]
> @Arcasias One note though, I don't know how to create two commits from the one I have. Any tips ?

@SimonGenin there are several options, but the one I usually use is to undo the current commit and add files/commit progressively. To undo a commit and keep its changes in the local diff use: `git reset HEAD~`

File: addons/web/static/src/js/fields/special_fields.js [hunk 11]
```diff
@@ -192,9 +194,71 @@ var FieldReportLayout = relational_fields.FieldMany2One.extend({
...
+
+    // If HTML, don't forget to adjust the sanitize options to avoid stripping most of the metadata
+    supportedFieldTypes: ['text', 'html'],
+
+    template: "web.IframeWrapper",
+
+    _render() {
+
+        const spinner = this.$el.find('.o_iframe_wrapper_spinner')[0];
+        const iframe = this.$el.find('.o_preview_iframe')[0];
```
Comment (Arcasias) [inline addons/web/static/src/js/fields/special_fields.js:208]: Just so you know, there is a shorthand on widget for this: `this.$('.o_iframe_wrapper_spinner')`. Now if you want to access native elements instead of jQuery I suggest you use `querySelector` on `el` (which is the native version of `$el`) instead, so:
```suggestion
        const spinner = this.el.querySelector('.o_iframe_wrapper_spinner');
        const iframe = this.el.querySelector('.o_preview_iframe');
```

File: addons/web/static/src/js/fields/special_fields.js [hunk 12]
```diff
@@ -192,9 +194,71 @@ var FieldReportLayout = relational_fields.FieldMany2One.extend({
...
+                 * model fields or with an rpc). It also cause other difficulties getting attribute on the most parent
+                 * nodes, parsing to HTML complex elements, etc.
+                 * Therefore, document.write makes it much more trivial in our situation.
+                 */
+                iframeDoc.open();
+                iframeDoc.write(htmlCode);
+                iframeDoc.close();
+
+                iframe.style.display = 'block';
+                spinner.style.display = 'none';
```
Comment (Arcasias) [inline addons/web/static/src/js/fields/special_fields.js:245]: I think you don't need the spinner beyond this point: might as well [remove it](https://developer.mozilla.org/en-US/docs/Web/API/ChildNode/remove).

File: addons/web/static/src/js/fields/special_fields.js [hunk 13]
```diff
@@ -192,9 +194,71 @@ var FieldReportLayout = relational_fields.FieldMany2One.extend({
...
+                 * others. In our case though, we deal with an iframe without src attribute and with metadata to put in
+                 * head tag. If we use the usual dom methods, the iframe is automatically created with its document
+                 * component containing html > head & body. Therefore, if we want to make it work that way, we would
+                 * need to receive each piece at a time to  append it to this document (with this.record.data and extra
+                 * model fields or with an rpc). It also cause other difficulties getting attribute on the most parent
+                 * nodes, parsing to HTML complex elements, etc.
+                 * Therefore, document.write makes it much more trivial in our situation.
+                 */
+                iframeDoc.open();
+                iframeDoc.write(htmlCode);
```
Comment (Arcasias) [inline addons/web/static/src/js/fields/special_fields.js:241]: Just noticed this now, but why not use `this.value` directly?

File: addons/web/models/base_document_layout.py [hunk 14]
```diff
@@ -181,9 +215,66 @@ def _parse_logo_colors(self, logo=None, white_threshold=225):
...
+        We give a useless company_ids arg, but provide the PREVIEW_ID arg that will prepare the template for
+        '_get_css_for_preview' processing later.
+        :return:
+        """
+        template_style = self.env.ref('web.styles_company_report', raise_if_not_found=False)
+        if not template_style:
+            return b''
+
+        company_styles = template_style._render({
+            'company_ids': [self],
```
Comment (kebeclibre) [inline addons/web/models/base_document_layout.py:237]: Sure about this?
self may be a record set with multiple elements.
so company_ids may be a list, with one iterable element, containing potentially multiple elements

File: addons/web/models/base_document_layout.py [hunk 15]
```diff
@@ -181,9 +215,66 @@ def _parse_logo_colors(self, logo=None, white_threshold=225):
...
+
+    @api.model
+    def _get_css_for_preview(self, scss, new_id):
+        """
+        Compile the scss into css. Since we had to get around company ids, the actual id has been replaced by
+        __PREVIEW_ID__. It get swapped by new_id, the company id used in the preview... Which is in reality the wizard
+        id. Which means it is not yet an id in DB, but a newId object.
+        """
+        css_code = self._compile_scss(scss)
+        css_code = css_code.replace('__PREVIEW_ID__', str(new_id))
```
Comment (kebeclibre) [inline addons/web/models/base_document_layout.py:251]: I don't like this because it forces very specific code in the template (a hack....)
I suppose in the template, company.id resolves to an object (a NewID in this case), and you want it to resolve to the function__str__ return ?
Isn't it possible to use already existing newID feature to discriminate, in the template, between
- id is really empty
- id is a NewID
- id is a integer

?

File: addons/web/static/src/js/fields/special_fields.js (hunk 13 above)
Comment (SimonGenin) [reply addons/web/static/src/js/fields/special_fields.js:241]: @Arcasias Self documented code :D But yeah, it can be this.value. It's because at first I didn't use the arrow function callback, so I needed to do a self = this. So it was the same amount of lines.

File: addons/web/static/src/js/fields/special_fields.js [hunk 16]
```diff
@@ -62,7 +64,7 @@ var FieldTimezoneMismatch = FieldSelection.extend({
     },
     /**
      * Display the timezone alert
-     * 
+     *
```
Comment (kebeclibre) [inline addons/web/static/src/js/fields/special_fields.js:67]: unneccasry diff

File: addons/web/static/src/js/fields/special_fields.js (hunk 13 above)
Comment (Arcasias) [reply addons/web/static/src/js/fields/special_fields.js:241]: Fair enough, that's a valid point :)

File: odoo/models.py [hunk 17]
```diff
@@ -204,6 +204,9 @@ def __repr__(self):
             "<NewId 0x%x>" % id(self)
         )
 
+    def __str__(self):
```
Comment (kebeclibre) [inline odoo/models.py:207]: I don't like this, linked to the previous "i don't like this" comment :D

File: addons/web/static/src/js/fields/special_fields.js [hunk 18]
```diff
@@ -62,7 +64,7 @@ var FieldTimezoneMismatch = FieldSelection.extend({
     },
     /**
      * Display the timezone alert
-     *
```
Comment (SimonGenin) [inline addons/web/static/src/js/fields/special_fields.js:65]: Capture d’écran 2020-08-19 à 11 50 01
Selon mon IDE, j'ai rien fait ici, je ne sais pas l'enlever...

File: addons/web/static/src/js/fields/special_fields.js (hunk 12 above)
Comment (SimonGenin) [reply addons/web/static/src/js/fields/special_fields.js:243]: @Arcasias Apparently, by using a template in the widget, the iframe and spinner are not recreated on every _render call. So remove it results in an error for the next onChange call.

Comment (Arcasias) [reply addons/web/static/src/js/fields/special_fields.js:243]: Alright, I guess I'm getting a bit too used to Owl and forgot how these worked ^^'

File: odoo/models.py [hunk 19]
```diff
@@ -204,6 +204,9 @@ def __repr__(self):
...
+        return "new_id_0x%x" % id(self)
```
Comment (kebeclibre) [inline odoo/models.py:208]: @rco-odoo 
Do you validate this. We need it to avoid having spaces in some css classes when doing `<t t-esc=company.id />` when company is a virtual record

Comment (kebeclibre) [reply odoo/models.py:208]: @Feyensv

Comment (rco-odoo) [reply odoo/models.py:208]: @kebeclibre how did you manage to inject new ids in templates?  That does not make sense.
Real programmers use real records :laughing:

Comment (Feyensv) [reply odoo/models.py:208]: > how did you manage to inject new ids in templates? That does not make sense.
> Real programmers use real records

If I'm not wrong, kanban can contain new records (in x2m fields) and support qweb.  It should probably happen in this case :thinking:

Comment (kebeclibre) [reply odoo/models.py:208]: @rco-odoo 
:D hahahaha
anywhooooo

The use case is:
- a wizard allows to modify fields on the company (there is not any real records in this case)
- some onchange allows to re-render the template of a report according to the wizard's values (there is no real records either.....)
- the values in db are irrelevant anyway (it is a transient model and the values of the company are not the ones we are manipulating)

Comment (rco-odoo) [reply odoo/models.py:208]: @kebeclibre okayyy, now it makes sense to me.
Have you tried to use `company._origin.id` instead of `company.id`?
`company._origin` is the real record corresponding to `company`, and is empty if there is no such record.
With that trick, you will always get a real id or `False`.

Comment (SimonGenin) [reply odoo/models.py:208]: @rco-odoo The new ids are here used *inside* css class name and html class names. And as explained above, we never have read ids since it's not yet store in database. But having False as result would not cut it.

**Comment by kebeclibre** [comment]
@rco-odoo 
@SimonGenin 
I modified the __str__ function of NewID to match what we discussed about

File: addons/web/data/report_layout.xml [hunk 20]
```diff
@@ -1,36 +1,41 @@
...
+
         <record id="report_layout_boxed" model="report.layout">
             <field name="name">Boxed</field>
             <field name="view_id" ref="web.external_layout_boxed"/>
             <field name="image">/web/static/img/preview_boxed.png</field>
             <field name="pdf">/web/static/pdf/preview_boxed.pdf</field>
         </record>
         <record id="report_layout_clean" model="report.layout">
-            <field name="name">Clean</field>
+            <field name="name">Bold</field>
```
Comment (kebeclibre) [inline addons/web/data/report_layout.xml:26]: are you sure about the changes ?

File: addons/web/data/report_layout.xml [hunk 21]
```diff
@@ -1,36 +1,41 @@
...
-        </record>
-        <record id="report_layout_background" model="report.layout">
+
+         <record id="report_layout_background" model="report.layout">
             <field name="name">Background</field>
             <field name="view_id" ref="web.external_layout_background"/>
             <field name="image">/web/static/img/preview_background.png</field>
             <field name="pdf">/web/static/pdf/preview_background.pdf</field>
         </record>
+
```
Comment (kebeclibre) [inline addons/web/data/report_layout.xml:11]: are you sure about the changes ?
those seem to just move around definitions of records

File: addons/web/models/base_document_layout.py [hunk 22]
```diff
@@ -33,9 +45,27 @@ class BaseDocumentLayout(models.TransientModel):
...
+                          sanitize=False,
+                          sanitize_tags=False,
+                          sanitize_attributes=False,
+                          sanitize_style=False,
+                          sanitize_form=False,
+                          strip_style=False,
+                          strip_classes=False)
+
+    # Those following fields are required as a company to create invoice report
+    partner_id = fields.Many2one(related='company_id.partner_id')
```
Comment (kebeclibre) [inline addons/web/models/base_document_layout.py:60]: fields in readonly

File: addons/web/views/report_templates.xml [hunk 23]
```diff
@@ -4,18 +4,13 @@
...
-            <!-- FIXME web_editor ? -->
           <link rel="stylesheet" type="text/scss" href="/web/static/src/scss/bootstrap_overridden_report.scss"/>
-          <link rel="stylesheet" type="text/scss" href="/web_editor/static/src/scss/bootstrap_overridden.scss"/>
         </t>
 
         <t t-call="web._assets_bootstrap"/>
 
         <link href="/base/static/src/css/description.css" rel="stylesheet" type="text/css"/>
 
-        <!-- FIXME web_editor ? -->
```
Comment (kebeclibre) [inline addons/web/views/report_templates.xml:16]: unnecessary changes, even kinda dangerous ??

**Review by kebeclibre — CHANGES_REQUESTED** [review]
There are some changes, removals and moving arounds that seem odd.....

File: addons/web/data/report_layout.xml (hunk 20 above)
Comment (SimonGenin) [reply addons/web/data/report_layout.xml:26]: @kebeclibre To be honest I think clean could be better than bold. The reason for the renaming of those is that it was wanted to set background (the default style) as first in the list. And I didn't like the fact that the style "Standart" was no longer the first (because it feels to me that if you call something standart, it should be displayed first as default). So the compromise with the PO was to give new names to avoid this line of thoughts. But if one is not of your liking we can change :)

File: addons/web/data/report_layout.xml (hunk 21 above)
Comment (SimonGenin) [reply addons/web/data/report_layout.xml:11]: This order defines the order in the view. Changes were requested so the default one would be first.

File: addons/web/models/base_document_layout.py (hunk 22 above)
Comment (SimonGenin) [reply addons/web/models/base_document_layout.py:60]: @kebeclibre And meant to be ? They are only used to retrieve info that could be required in the invoice template, they are never changes from the wizard.

File: addons/web/views/report_templates.xml (hunk 23 above)
Comment (SimonGenin) [reply addons/web/views/report_templates.xml:16]: @kebeclibre Well, it kinda is yes. Problem is, with those lines, there is a scss problem in tests. It says that a variable is missing (which is true, the files here requires data from other non-imported web editor files). And those lines are called again in the webeditor.xml. I mentionned it here in the discord: https://discordapp.com/channels/678381219515465750/687338039717920792/742996732450963516 and following the answers I tried to remove it. No test break so technically we're safe :D (just kidding of course)

Comment (SimonGenin) [reply addons/web/views/report_templates.xml:16]: Ho, and it's not one test that fails. It's more like for every tests, it screams there is a problem in the scss compilation because of a missing var

File: addons/web/data/report_layout.xml (hunk 21 above)
Comment (kebeclibre) [reply addons/web/data/report_layout.xml:11]: @SimonGenin 
hmmmmm, this is fairly weak. I'd would have preferred a new field "sequence" on report.layout

File: addons/web/models/base_document_layout.py (hunk 22 above)
Comment (kebeclibre) [reply addons/web/models/base_document_layout.py:60]: yes I know, It is both for clarity in the code and some sense of security... One should never be able to change those data from a simple wizard

File: addons/web/views/report_templates.xml (hunk 23 above)
Comment (kebeclibre) [reply addons/web/views/report_templates.xml:16]: fair enough :D

File: addons/web/models/base_document_layout.py (hunk 22 above)
Comment (SimonGenin) [reply addons/web/models/base_document_layout.py:60]: oh you're talking about the preview field, I though you were talking about the related fields below.

File: addons/web/data/report_layout.xml (hunk 21 above)
Comment (SimonGenin) [reply addons/web/data/report_layout.xml:11]: I can look into that

File: addons/web/models/base_document_layout.py (hunk 22 above)
Comment (kebeclibre) [reply addons/web/models/base_document_layout.py:60]: héhéhé, nope, I'm definitely talking about the related fields (partner, vat etc.....)
Their presence in code is just to provide a res.company-like API, they should never be writable from the wizard

**Comment by kebeclibre** [comment]
retargetted saas-13.5

File: addons/web/models/base_document_layout.py (hunk 22 above)
Comment (SimonGenin) [reply addons/web/models/base_document_layout.py:60]: I though related field were readonly by default ? You're saying you want it to be more explicit ?

File: odoo/release.py [hunk 24]
```diff
@@ -12,7 +12,7 @@
 # properly comparable using normal operarors, for example:
 #  (6,1,0,'beta',0) < (6,1,0,'candidate',1) < (6,1,0,'candidate',2)
 #  (6,1,0,'candidate',2) < (6,1,0,'final',0) < (6,1,2,'final',0)
-version_info = (13, 5, 0, ALPHA, 1, '')
+version_info = ('saas~13', 5, 0, FINAL, 0, '')
```
Comment (SimonGenin) [inline odoo/release.py:15]: @kebeclibre Is this how we do it ?

Comment (kebeclibre) [reply odoo/release.py:15]: @SimonGenin 
Yes, this is how a relase is made, (that and the fork to its own branch)

You'll need to cherry-pick your commit on top of a fresh saas-13.5 (and redo a PR for that release)

**Comment by SimonGenin** [comment]
New PR for target 13.5 https://github.com/odoo/odoo/pull/56415

**Comment by kebeclibre** [comment]
migration https://github.com/odoo/upgrade/pull/1686
